nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/nar/NarRestApiClient.java

```diff
@@ -0,0 +1,138 @@
+                .port(baseUri.getPort())
+                .addPathSegment(NIFI_API_PATH)
+                .addPathSegment(CONTROLLER_PATH)
+                .addPathSegment(NAR_MANAGER_PATH)
+                .addPathSegment(NARS_PATH)
+                .build();
+        LOGGER.info("Requesting NAR summaries from {}", requestUri);
+
+        final HttpRequestBodySpec requestBodySpec = webClientService.get()
+                .uri(requestUri)
+                .header("Accept", "application/json");
+
+        try (final HttpResponseEntity response = requestBodySpec.retrieve()) {
+            final InputStream responseBody = getResponseBody(requestUri, response);
+            return objectMapper.readValue(responseBody, NarSummariesEntity.class);
+        } catch (final WebClientServiceException | IOException e) {
+            throw new NarRestApiRetryableException(e.getMessage(), e);
+        }
+    }
+
+    public InputStream downloadNar(final String identifier) {
+        final URI requestUri = new StandardHttpUriBuilder()
+                .scheme(baseUri.getScheme())
+                .host(baseUri.getHost())
+                .port(baseUri.getPort())
+                .addPathSegment(NIFI_API_PATH)
+                .addPathSegment(CONTROLLER_PATH)
+                .addPathSegment(NAR_MANAGER_PATH)
+                .addPathSegment(NARS_PATH)
+                .addPathSegment(identifier)
+                .addPathSegment(NAR_DOWNLOAD_PATH)
+                .build();
+        LOGGER.info("Downloading NAR [{}] from {}", identifier, requestUri);
+
+        try {
+            final HttpResponseEntity response = webClientService.get()
+                    .uri(requestUri)
+                    .header("Accept", "application/octet-stream")
+                    .retrieve();
+            return getResponseBody(requestUri, response);
+        } catch (final WebClientServiceException e) {
+            throw new NarRestApiRetryableException(e.getMessage(), e);
+        }
+    }
+
+    private InputStream getResponseBody(final URI requestUri, final HttpResponseEntity response) {
+        final int statusCode = response.statusCode();
+        if (HttpResponseStatus.isSuccessful(statusCode)) {
+            return response.body();
+        } else {
+            final String responseMessage;
+            try {
+                responseMessage = IOUtils.toString(response.body(), StandardCharsets.UTF_8);
+            } catch (IOException e) {
+                throw new NarRestApiRetryableException("Error reading response from " + requestUri + " - " + statusCode, e);
+            }
+            if (statusCode == HttpResponseStatus.CONFLICT.getCode()) {
+                throw new NarRestApiRetryableException("Error calling " + requestUri + " - " + statusCode + " - " + responseMessage);
+            } else {
+                throw new IllegalStateException("Error calling " + requestUri + " - " + statusCode + " - " + responseMessage);
+            }
```

Comment:
Recommend using `.formatted()` to construct the error message strings for readability.